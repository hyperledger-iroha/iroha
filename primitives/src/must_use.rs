//! Contains wrapper type to annotate types with `must_use` attribute

use core::borrow::{Borrow, BorrowMut};

use derive_more::{AsMut, AsRef, Constructor, Deref, Display};

/// Wrapper type to annotate types with `must_use` attribute
///
/// # Example
/// ```
/// use iroha_primitives::must_use::MustUse;
///
/// fn is_odd(num: i32) -> Result<MustUse<bool>, String> {
///     if num < 0 {
///         return Err(String::from("Number must be positive"));
///     }
///
///     if num % 2 == 0 {
///         Ok(MustUse::new(true))
///     } else {
///         Ok(MustUse::new(false))
///     }
/// }
///
/// if *is_odd(2).unwrap() {
///     println!("2 is odd");
/// }
///
/// // Will produce a warning, case `#[warn(unused_must_use)]` on by default
/// // is_odd(3).unwrap();
/// ```
#[derive(
    Constructor,
    Debug,
    Display,
    Copy,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    AsRef,
    AsMut,
    Deref,
)]
#[repr(transparent)]
#[must_use]
pub struct MustUse<T>(pub T);

impl<T> MustUse<T> {
    /// Get inner value
    #[inline]
    pub fn into_inner(self) -> T {
        self.0
    }
}

impl<T> Borrow<T> for MustUse<T> {
    #[inline]
    fn borrow(&self) -> &T {
        &self.0
    }
}

impl<T> BorrowMut<T> for MustUse<T> {
    #[inline]
    fn borrow_mut(&mut self) -> &mut T {
        &mut self.0
    }
}
