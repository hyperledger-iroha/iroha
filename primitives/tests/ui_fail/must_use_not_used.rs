#![deny(unused_must_use)]

use iroha_primitives::must_use::MustUse;

fn main() {
    MustUse::new(5);
}
