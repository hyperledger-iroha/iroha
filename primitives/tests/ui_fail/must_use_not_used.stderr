error: unused `MustUse` that must be used
 --> tests/ui_fail/must_use_not_used.rs:6:5
  |
6 |     MustUse::new(5);
  |     ^^^^^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> tests/ui_fail/must_use_not_used.rs:1:9
  |
1 | #![deny(unused_must_use)]
  |         ^^^^^^^^^^^^^^^
